FWCore/Utilities/interface/ReusableObjectHolder.h

```diff
@@ -108,43 +108,44 @@ namespace edm {
     /// Use this function in conjunction with add()
     std::shared_ptr<T> tryToGet() {
       std::unique_ptr<T, Deleter> item;
-      m_availableQueue.try_pop(item);
-      if (nullptr == item) {
+      if (m_availableQueue.try_pop(item)) {
+        return wrapCustomDeleter(std::move(item));
+      } else {
         return std::shared_ptr<T>{};
       }
-      //instead of deleting, hand back to queue
-      auto pHolder = this;
-      auto deleter = item.get_deleter();
-      ++m_outstandingObjects;
-      return std::shared_ptr<T>{item.release(), [pHolder, deleter](T* iItem) {
-                                  pHolder->addBack(std::unique_ptr<T, Deleter>{iItem, deleter});
-                                }};
     }
 
-    ///If there isn't an object already available, creates a new one using iFunc
+    ///Takes an object from the queue if one is available, or creates one using iFunc.
     template <typename F>
     std::shared_ptr<T> makeOrGet(F iFunc) {
-      std::shared_ptr<T> returnValue;
-      while (!(returnValue = tryToGet())) {
-        add(makeUnique(iFunc()));
+      std::unique_ptr<T, Deleter> item;
+      if (m_availableQueue.try_pop(item)) {
+        return wrapCustomDeleter(std::move(item));
+      } else {
+        return wrapCustomDeleter(makeUnique(iFunc()));
       }
-      return returnValue;
     }
 
-    ///If there is an object already available, passes the object to iClearFunc and then
-    /// returns the object.
-    ///If there is not an object already available, creates a new one using iMakeFunc
+    ///Takes an object from the queue if one is available, or creates one using iMakeFunc.
+    ///Then, passes the object to iClearFunc, and returns it.
     template <typename FM, typename FC>
     std::shared_ptr<T> makeOrGetAndClear(FM iMakeFunc, FC iClearFunc) {
-      std::shared_ptr<T> returnValue;
-      while (!(returnValue = tryToGet())) {
-        add(makeUnique(iMakeFunc()));
-      }
+      std::shared_ptr<T> returnValue = makeOrGet(iMakeFunc);
```

Comment:
I believe doing
```suggestion
      std::shared_ptr<T> returnValue = makeOrGet(std::move(iMakeFunc));
```
would avoid a possible copy of the functor.

Possibly an even better solution would be to change the function signature to
```C++
makeOrGetAndClear(FM&& iMakeFunc, FC&& iClearFunc)
```

and then do
```suggestion
      std::shared_ptr<T> returnValue = makeOrGet(std::forward<FM>(iMakeFunc));
```
A similar function signature change to `makeOrGet` would probably be required to get the most benefit of such a change.

Reply:
Good catch, thanks.

Reply:
I think this change
```c++
    template <typename FM, typename FC>
    std::shared_ptr<T> makeOrGetAndClear(FM&& iMakeFunc, FC iClearFunc) {
      std::shared_ptr<T> returnValue = makeOrGet(std::forward<FM>(iMakeFunc));
      iClearFunc(returnValue.get());
      return returnValue;
    }
```
should be enough, since `iMakeFunc` is the only one that is copied.

Reply:
At least, my confused understanding is that `FC iClearFunc` (in `makeOrGetAndClear`) and `F iFunc` (in `makeOrGet`) should already any of the functors by reference if they are lvalues.

Reply:
I think using `FC` instead of `FC&&` means the compiler (before optimization) will call the move constructor of `FC` when passed an lvalue. I'd bet that all compilers will remove that call when optimation is turned on. However, using `FC&&` does make the requirements of the arguments more explicit.

Reply:
I was wrong. I tested in godbolt and for the case where the object is created directly in the argument list they are identical.

Reply:
It does appear that if the argument is created outside of the function argument list, then `FC&&` is the better choice as it does avoid an extra copy.

https://godbolt.org/z/7fYG641qq

Reply:
OK, done - let me know if I missed anything (here or elsewhere).

Reply:
Sorry, I missed the last comment...

Reply:
...done, now.